File: src/waypalm-site/src/main/java/waypalm/site/validation/WorldTitleValidator.java
package waypalm.site.validation;

import org.springframework.beans.BeanUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.transaction.annotation.Transactional;
import waypalm.domain.dao.WorldRepository;
import waypalm.domain.entity.World;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;
import java.beans.PropertyDescriptor;

public class WorldTitleValidator implements ConstraintValidator<WorldTitle, Object> {
    @Autowired
    protected WorldRepository worldRepository;

    private String idField;
    private String titleField;

    @Override
    public void initialize(WorldTitle constraintAnnotation) {
        idField = constraintAnnotation.idField();
        titleField = constraintAnnotation.titleField();
    }

    @Override
    @Transactional
    public boolean isValid(Object object, ConstraintValidatorContext context) {
        if (object == null) {
            return false;
        }

        String titleValue = readTitle(object);
        Integer idValue = readId(object);

        World world = worldRepository.findWorldByTitle(titleValue);
        if (world == null || world.getId().equals(idValue)) {
            return true;
        }

        context.buildConstraintViolationWithTemplate(context.getDefaultConstraintMessageTemplate())
                .addNode(titleField)
                .addConstraintViolation()
                .disableDefaultConstraintViolation();
        return false;
    }

    private Integer readId(Object object) {
        if (idField.length() == 0) {
            return null;
        }

        PropertyDescriptor idProperty = BeanUtils.getPropertyDescriptor(object.getClass(), idField);
        if (idProperty == null) {
            throw new IllegalStateException("property set is not complete");
        }
        try {
            return (Integer) idProperty.getReadMethod().invoke(object);
        } catch (Exception e) {
            throw new IllegalStateException("property set is not readable");
        }
    }

    private String readTitle(Object object) {
        PropertyDescriptor titleProperty = BeanUtils.getPropertyDescriptor(object.getClass(), titleField);
        if (titleProperty == null) {
            throw new IllegalStateException("property set is not complete");
        }
        try {
            return (String) titleProperty.getReadMethod().invoke(object);
        } catch (Exception e) {
            throw new IllegalStateException("property set is not readable");
        }
    }
}

File: src/waypalm-site/src/main/java/waypalm/site/validation/WorldTitleValidator.kt
package waypalm.site.validation

import javax.validation.ConstraintValidator
import javax.inject.Inject
import waypalm.domain.dao.WorldRepository
import javax.validation.ConstraintValidatorContext
import org.springframework.beans.BeanUtils

public class WorldTitleValidator [Inject] (
        val  worldRepository: WorldRepository,
        var idField: String = "",
        var titleField: String = ""
)
: ConstraintValidator<WorldTitle, Any>
{
    public override fun initialize(constraintAnnotation: WorldTitle?) {
        idField = constraintAnnotation!!.idField() as String;
        titleField = constraintAnnotation.titleField() as String;
    }

    public override fun isValid(value: Any?, context: ConstraintValidatorContext?): Boolean {
        if (value == null) {
            return false;
        }

        val titleValue = readTitle(value);
        val idValue = readId(value);

        val world = worldRepository.findWorldByTitle(titleValue);
        if (world == null || world.getId().equals(idValue)) {
            return true;
        }

        context!!
                .buildConstraintViolationWithTemplate(context.getDefaultConstraintMessageTemplate())!!
                .addNode(titleField)!!
                .addConstraintViolation()!!
                .disableDefaultConstraintViolation();
        return false;
    }


    private fun readId(obj: Any): Int? {
        if (idField.length() == 0) {
            return null;
        }

        val idProperty = BeanUtils.getPropertyDescriptor(obj.javaClass, idField);
        if (idProperty == null) {
            throw IllegalStateException("property set is not complete");
        }
        try {
            return idProperty.getReadMethod()!!.invoke(obj) as Int;
        } catch (e: Exception) {
            throw IllegalStateException("property set is not readable");
        }
    }

    private fun readTitle(obj: Any): String {
        val titleProperty = BeanUtils.getPropertyDescriptor(obj.javaClass, titleField);
        if (titleProperty == null) {
            throw  IllegalStateException("property set is not complete");
        }
        try {
            return  titleProperty.getReadMethod()!!.invoke(obj) as String;
        } catch (e: Exception) {
            throw IllegalStateException("property set is not readable");
        }
    }
}